## 4.6.2 Combining & Exporting Data

### This script contains the following points:
#### 1. Import libraries and Data sets into Jupyter
#### 2. Check dimension for imported combined dataframes
#### 3. Determine a suitable way to combine the orders_products_combined dataframe with the products data set
#### 4. Checking the counts of the result using the merge flag
#### 5. Export the newly created dataframe as ords_prods_merge in a pkl format

### 1.Import libraries and Data sets into Jupyter

In [2]:

import pandas as pd
import numpy as np
import os

In [3]:

path = r'E:\Aarthi Sarathi\Career Foundry\Exercise 4\Instacart Basket Analsis'
df_ords_prods_combined = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.csv'))
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

### 2.Check dimension for imported combined dataframes

In [4]:
df_ords_prods_combined.head()

Unnamed: 0  order_id  product_id  add_to_cart_order  reordered  user_id  \
0           0         2       33120                  1          1   202279   
1           1         2       28985                  2          1   202279   
2           2         2        9327                  3          0   202279   
3           3         2       45918                  4          1   202279   
4           4         2       30035                  5          0   202279   

  eval_set  order_number  order_dow  order_hour_of_day  \
0    prior             3          5                  9   
1    prior             3          5                  9   
2    prior             3          5                  9   
3    prior             3          5                  9   
4    prior             3          5                  9   

   days_since_prior_order _merge  
0                     8.0   both  
1                     8.0   both  
2                     8.0   both  
3                     8.0   both  
4                     8.0   both

In [5]:
df_ords_prods_combined.shape

(32434489, 12)

In [6]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [7]:
#Dropping unnecessary columns(as unnamed)

df_prods = df_prods.drop(['Unnamed: 0'], axis=1)
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [8]:
df_prods.shape

(49672, 5)

### 3.Determine a suitable way to combine the orders_products_combined dataframe with your products data set

In [9]:
#As this df_ords_prods_combined already has _merge column so dropping that

df_ords_prods_combined = df_ords_prods_combined.drop(['_merge'], axis=1)
#check output
df_ords_prods_combined.head()

Unnamed: 0  order_id  product_id  add_to_cart_order  reordered  user_id  \
0           0         2       33120                  1          1   202279   
1           1         2       28985                  2          1   202279   
2           2         2        9327                  3          0   202279   
3           3         2       45918                  4          1   202279   
4           4         2       30035                  5          0   202279   

  eval_set  order_number  order_dow  order_hour_of_day  days_since_prior_order  
0    prior             3          5                  9                     8.0  
1    prior             3          5                  9                     8.0  
2    prior             3          5                  9                     8.0  
3    prior             3          5                  9                     8.0  
4    prior             3          5                  9                     8.0

In [10]:
df_ords_prods_combined['Unnamed: 0']

0                  0
1                  1
2                  2
3                  3
4                  4
              ...   
32434484    32434484
32434485    32434485
32434486    32434486
32434487    32434487
32434488    32434488
Name: Unnamed: 0, Length: 32434489, dtype: int64

## I have optimized data to avoid memory error

In [11]:
df_ords_prods_combined['order_id']=df_ords_prods_combined['order_id'].astype('int32')
df_ords_prods_combined['product_id']=df_ords_prods_combined['product_id'].astype('int32')
df_ords_prods_combined['add_to_cart_order']=df_ords_prods_combined['add_to_cart_order'].astype('int32')
df_ords_prods_combined['reordered']=df_ords_prods_combined['reordered'].astype('int32')
df_ords_prods_combined['order_number']=df_ords_prods_combined['order_number'].astype('int32')
df_ords_prods_combined['order_dow']=df_ords_prods_combined['order_dow'].astype('int32')
df_ords_prods_combined['order_dow']=df_ords_prods_combined['order_dow'].astype('int32')
df_ords_prods_combined['order_hour_of_day']=df_ords_prods_combined['order_hour_of_day'].astype('int32')
df_ords_prods_combined['days_since_prior_order']=df_ords_prods_combined['days_since_prior_order'].astype('float32')
df_ords_prods_combined['user_id']=df_ords_prods_combined['user_id'].astype('int32')

In [12]:
df_ords_prods_combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 11 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int32  
 2   product_id              int32  
 3   add_to_cart_order       int32  
 4   reordered               int32  
 5   user_id                 int32  
 6   eval_set                object 
 7   order_number            int32  
 8   order_dow               int32  
 9   order_hour_of_day       int32  
 10  days_since_prior_order  float32
dtypes: float32(1), int32(8), int64(1), object(1)
memory usage: 1.6+ GB


In [13]:
df_ords_prods_combined.head()

Unnamed: 0  order_id  product_id  add_to_cart_order  reordered  user_id  \
0           0         2       33120                  1          1   202279   
1           1         2       28985                  2          1   202279   
2           2         2        9327                  3          0   202279   
3           3         2       45918                  4          1   202279   
4           4         2       30035                  5          0   202279   

  eval_set  order_number  order_dow  order_hour_of_day  days_since_prior_order  
0    prior             3          5                  9                     8.0  
1    prior             3          5                  9                     8.0  
2    prior             3          5                  9                     8.0  
3    prior             3          5                  9                     8.0  
4    prior             3          5                  9                     8.0

In [14]:
df_ords_prods_combined.drop(['Unnamed: 0'], axis=1)

order_id  product_id  add_to_cart_order  reordered  user_id  \
0                2       33120                  1          1   202279   
1                2       28985                  2          1   202279   
2                2        9327                  3          0   202279   
3                2       45918                  4          1   202279   
4                2       30035                  5          0   202279   
...            ...         ...                ...        ...      ...   
32434484   3421083       39678                  6          1    25247   
32434485   3421083       11352                  7          0    25247   
32434486   3421083        4600                  8          0    25247   
32434487   3421083       24852                  9          1    25247   
32434488   3421083        5020                 10          1    25247   

         eval_set  order_number  order_dow  order_hour_of_day  \
0           prior             3          5                  9   
1           prior             3          5                  9   
2           prior             3          5                  9   
3           prior             3          5                  9   
4           prior             3          5                  9   
...           ...           ...        ...                ...   
32434484    prior            24          2                  6   
32434485    prior            24          2                  6   
32434486    prior            24          2                  6   
32434487    prior            24          2                  6   
32434488    prior            24          2                  6   

          days_since_prior_order  
0                            8.0  
1                            8.0  
2                            8.0  
3                            8.0  
4                            8.0  
...                          ...  
32434484                    21.0  
32434485                    21.0  
32434486                    21.0  
32434487                    21.0  
32434488                    21.0  

[32434489 rows x 10 columns]

In [15]:
df_ords_prods_combined.dtypes

Unnamed: 0                  int64
order_id                    int32
product_id                  int32
add_to_cart_order           int32
reordered                   int32
user_id                     int32
eval_set                   object
order_number                int32
order_dow                   int32
order_hour_of_day           int32
days_since_prior_order    float32
dtype: object

#### Merging df_ords_prods_combined with df_prods

In [16]:
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined , on = 'product_id' , indicator = True)
#check output
df_ords_prods_merge

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
32404854       49688      Fresh Foaming Cleanser        73             11   
32404855       49688      Fresh Foaming Cleanser        73             11   
32404856       49688      Fresh Foaming Cleanser        73             11   
32404857       49688      Fresh Foaming Cleanser        73             11   
32404858       49688      Fresh Foaming Cleanser        73             11   

          prices  Unnamed: 0  order_id  add_to_cart_order  reordered  user_id  \
0            5.8       10708      1107                  7          0    38259   
1            5.8       50379      5319                  3          1   196224   
2            5.8       70912      7540                  4          1   138499   
3            5.8       86636      9228                  2          0    79603   
4            5.8       87149      9273                 30          0    50005   
...          ...         ...       ...                ...        ...      ...   
32404854    13.5    29506595   3111954                  3          0    79411   
32404855    13.5    29602387   3122003                 20          0   146323   
32404856    13.5    30023199   3166828                  5          1   188073   
32404857    13.5    31194081   3290206                 10          0   121646   
32404858    13.5    32246888   3401313                  5          0   200377   

         eval_set  order_number  order_dow  order_hour_of_day  \
0           prior             2          1                 11   
1           prior            65          1                 14   
2           prior             8          0                 14   
3           prior             2          2                 10   
4           prior             1          1                 15   
...           ...           ...        ...                ...   
32404854    prior             6          4                 19   
32404855    prior             3          1                 16   
32404856    prior             7          1                 15   
32404857    prior             3          0                  9   
32404858    prior             1          4                 11   

          days_since_prior_order _merge  
0                            7.0   both  
1                            1.0   both  
2                            7.0   both  
3                           30.0   both  
4                            NaN   both  
...                          ...    ...  
32404854                     4.0   both  
32404855                    24.0   both  
32404856                    30.0   both  
32404857                    14.0   both  
32404858                     NaN   both  

[32404859 rows x 16 columns]

### 4.Checking the counts of the result using the merge flag

In [17]:
#Checking counts in merged dataframe
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### The resulting dataframe (after the merge) has 32,404,859 rows, and each of those rows have information found in both input data sets, as we used an inner join for the purposes of this project.

### 5.Export the newly created dataframe as ords_prods_merge in a pkl format

In [18]:
df_ords_prods_merge.rename(columns={'order_dow': 'order_date_of_the_week'}, inplace=True)


In [22]:
df_ords_prods_merge = df_ords_prods_merge.drop(['Unnamed: 0'], axis=1)

In [23]:
df_ords_prods_merge = df_ords_prods_merge.drop(['eval_set'], axis=1)

df_ords_prods_merge.head()

In [25]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))
df_ords_prods_merge.to_csv(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.csv'))